components/webln/nwc.js

```diff
@@ -0,0 +1,247 @@
+// https://github.com/getAlby/js-sdk/blob/master/src/webln/NostrWeblnProvider.ts
+
+import { createContext, useCallback, useContext, useEffect, useState } from 'react'
+import { Relay, finalizeEvent, nip04 } from 'nostr-tools'
+
+const NWCContext = createContext()
+
+export function NWCProvider ({ children }) {
+  const [nwcUrl, setNwcUrl] = useState('')
+  const [walletPubkey, setWalletPubkey] = useState()
+  const [relayUrl, setRelayUrl] = useState()
+  const [secret, setSecret] = useState()
+  const [enabled, setEnabled] = useState()
+  const [relay, setRelay] = useState()
+
+  const name = 'NWC'
+  const storageKey = 'webln:provider:nwc'
+
+  const loadConfig = useCallback(async () => {
+    const config = window.localStorage.getItem(storageKey)
+    if (!config) return
+
+    const configJSON = JSON.parse(config)
+
+    const { nwcUrl } = configJSON
+    setNwcUrl(nwcUrl)
+    if (!nwcUrl) {
+      setEnabled(undefined)
+      return
+    }
+
+    const params = parseWalletConnectUrl(nwcUrl)
+    setRelayUrl(params.relayUrl)
+    setWalletPubkey(params.walletPubkey)
+    setSecret(params.secret)
+
+    try {
+      const supported = await validateParams(params)
+      setEnabled(supported.includes('pay_invoice'))
+    } catch (err) {
+      console.error('invalid NWC config:', err)
+      setEnabled(false)
+      throw err
+    }
+  }, [])
+
+  const saveConfig = useCallback(async (config) => {
+    // immediately store config so it's not lost even if config is invalid
+    const { nwcUrl } = config
+    setNwcUrl(nwcUrl)
+    if (!nwcUrl) {
+      setEnabled(undefined)
+      return
+    }
+
+    const params = parseWalletConnectUrl(nwcUrl)
+    setRelayUrl(params.relayUrl)
+    setWalletPubkey(params.walletPubkey)
+    setSecret(params.secret)
+
+    // XXX Even though NWC allows to configure budget,
+    // this is definitely not ideal from a security perspective.
+    window.localStorage.setItem(storageKey, JSON.stringify(config))
+
+    try {
+      const supported = await validateParams(params)
+      setEnabled(supported.includes('pay_invoice'))
+    } catch (err) {
+      console.error('invalid NWC config:', err)
+      setEnabled(false)
+      throw err
+    }
+  }, [])
+
+  const clearConfig = useCallback(() => {
+    window.localStorage.removeItem(storageKey)
+    setNwcUrl('')
+    setRelayUrl(undefined)
+    setWalletPubkey(undefined)
+    setSecret(undefined)
+    setEnabled(undefined)
+  }, [])
+
+  useEffect(() => {
```

Comment:
Do we need the `useEffect` here? Just double checking because you managed to remove the similar logic in the lnbits provider.

Reply:
I could also add logic to `loadConfig` and `saveConfig` to open/close relays on NWC url changes. But I think this `useEffect` here is actually pretty nice since it handles closing the relay automatically with the cleanup function and the dependency also automatically handles not closing the relay if the relay URL did not change on a save.

Looking at `loadConfig` and `saveConfig` again, I think they could be refactored even more (since they still have some shared code) and handling the relay connection there would make it more complicated.

So I think this `useEffect` is doing us some favors.